### PR TITLE
 feat(injector): add `component` to `shouldInclude`

### PR DESCRIPTION
**Pull request by eps1lon**
Edit: ~I think I have a better idea that solves the same issue: Let's include the filename where the prop/component symnol is declared. That way we can leverage `includeUnusedProps` and filter by filename in `shouldInclude`. In the end we'll probably filter out any prop type that's not declared in the same file as the component.~ It's still needed because `includeUnusedProps` causes `usedProps` to be empty.

Two goals are achieved with this change:
- `includeUnusedProps: true` + `shouldInclude` as a ignore list doesn't scale very well if you extend very large prop types (e.g. HTML Attributes). 
  It seemed like it bailed out early if `includeUnusedProps` was `true` even though we might have included it later via `shouldInclude`
- we want to move from custom `@document` directives to internal allow lists. Having magic markers in the jsdoc means that library users also see these magic markers.

I've separated this PR into test and implementation in case there's a better implementation (or even existing options) to achieve the same goal.

### CODE REVIEW

File: src/injector.ts [hunk 1]
```diff
@@ -156,15 +160,11 @@ function plugin(
 
 				let usedProps: string[] = [];
 
-				if (!includeUnusedProps) {
```
Comment (eps1lon) [inline src/injector.ts]: Just because I don't want to include unused props by default doesn't meant I might actually want to include them via `shouldInclude`.

Comment (merceyz) [reply src/injector.ts]: Fixed in https://github.com/merceyz/typescript-to-proptypes/commit/277258ddc73c3da816aba6fccb739c69dfe8e83a

File: src/generator.ts [hunk 2]
```diff
@@ -34,6 +34,8 @@ export interface GenerateOptions {
 	 * }
 	 */
 	comment?: string;
+
+	component?: t.ComponentNode;
```
Comment (merceyz) [inline src/generator.ts]: Could you add some documentation please

File: test/injector/should-include-component-based/input.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+	 */
+	onChange?: () => any;
+}
+
+export function Snackbar(props: SnackBarProps) {
+	return <div {...props} />;
+}
+
+// here we don't care about `key`
+export function SomeOtherComponent(props: { children?: React.ReactNode }) {
```
Comment (merceyz) [inline test/injector/should-include-component-based/input.tsx]: Should this be changed to use `SnackBarProps`? Otherwise it seems unrelated

File: src/generator.ts [hunk 4]
```diff
@@ -67,7 +70,8 @@ export function generate(node: t.Node | t.PropTypeNode[], options: GenerateOptio
 
 		let filteredNodes = node;
 		if (shouldInclude) {
-			filteredNodes = filteredNodes.filter((x) => shouldInclude(x));
+			// we're in a component context so it's definitely not-null
```
Comment (merceyz) [inline src/generator.ts]: Not necessarily, generate could be called directly with `PropTypeNode[]`, in which case there is no component.
The assumption is however correct if it's only called by the injector

File: src/generator.ts [hunk 5]
```diff
@@ -24,7 +24,7 @@ export interface GenerateOptions {
 	 * Control which PropTypes are included in the final result
 	 * @param proptype The current PropType about to be converted to text
 	 */
-	shouldInclude?(proptype: t.PropTypeNode): boolean | undefined;
+	shouldInclude?(proptype: t.PropTypeNode, component: t.ComponentNode): boolean | undefined;
```
Comment (merceyz) [inline src/generator.ts]:
```suggestion
	shouldInclude?(proptype: t.PropTypeNode, component?: t.ComponentNode): boolean | undefined;
```

Could you also document this param please

File: src/injector.ts [hunk 6]
```diff
@@ -21,7 +21,11 @@ export type InjectOptions = {
 	 * use the default behaviour
 	 * @default includeUnusedProps ? true : data.usedProps.includes(data.prop.name)
 	 */
-	shouldInclude?(data: { prop: t.PropTypeNode; usedProps: string[] }): boolean | undefined;
+	shouldInclude?(data: {
+		component: t.ComponentNode;
```
Comment (merceyz) [inline src/injector.ts:25]:
```suggestion
		component?: t.ComponentNode;
```

**Review by merceyz — CHANGES_REQUESTED** [review]
Left some feedback inline, otherwise it looks good

> It seemed like it bailed out early if includeUnusedProps was true even though we might have included it later via shouldInclude
>It's still needed because includeUnusedProps causes usedProps to be empty.

Fixed in https://github.com/merceyz/typescript-to-proptypes/commit/277258ddc73c3da816aba6fccb739c69dfe8e83a since it required some more work and as it is unrelated to this PR.

File: test/injector/should-include-component-based/input.tsx (hunk 3 above)
Comment (eps1lon) [reply test/injector/should-include-component-based/input.tsx]: It shows how we care about `key` in one component but not the other. Without `component` we wouldn't be able to differentiate (unless we would match the jsdoc which is not robust).

Comment (merceyz) [reply test/injector/should-include-component-based/input.tsx]: Yeah, but the component doesn't have a `key` prop. Which is why I think you meant to have it use `SnackBarProps`

Comment (eps1lon) [reply test/injector/should-include-component-based/input.tsx]: Do you infer the props that can be passed to `createElement` or short-circuit type inference and go directly to the props interface? Because every component has a `key` prop implicitly.

Comment (merceyz) [reply test/injector/should-include-component-based/input.tsx]: It only parses declared types, so as you can see from `output.json`, `SomeOtherComponent` doesn't have a prop named `key`

Comment (eps1lon) [reply test/injector/should-include-component-based/input.tsx]: Well then let's stick with children. Either way this should not use SnackBarProps.

Comment (eps1lon) [reply test/injector/should-include-component-based/input.tsx]: done

File: src/generator.ts (hunk 5 above)
Comment (eps1lon) [reply src/generator.ts]: Why would it be `undefined` here?

Comment (eps1lon) [reply src/generator.ts]: The other params aren't document either. I don't think there's anything meaningful to add. It will probably boil down to "the component of the prop" which is just repeating the code.

File: src/generator.ts (hunk 2 above)
Comment (eps1lon) [reply src/generator.ts]: done

File: src/generator.ts (hunk 5 above)
Comment (eps1lon) [reply src/generator.ts]: nvm wrong place, done

**Review by merceyz — APPROVED** [review]

**Comment by merceyz** [comment]
Thanks, the failing build seems to be yarn generating different hashes on Windows and Linux
